crates/egui_demo_app/src/apps/custom3d_wgpu.rs

```diff
@@ -135,12 +135,18 @@ impl Custom3d {
         // Device and Queue, which can be used, for instance, to update buffers and uniforms before
         // rendering.
         //
+        // You can use the main `CommandEncoder` that is passed-in, return an arbitrary number
+        // of user-defined `CommandBuffer`s, or both.
```

Comment:
Why the two ways of doing it? Are the `CommandEncoder` expensive to create, so better to reuse the same one for the common case?

Reply:
correct. Although there is some lack of clarity how expensive exactly

Reply:
the only thing that is very clear is that submit calls are very expensive - that's even in the official wgpu advice
https://github.com/gfx-rs/wgpu/wiki/Do's-and-Dont's#dont-submit-many-times-per-frame

Reply:
Actually https://developer.nvidia.com/blog/vulkan-dos-donts/ is talking about 5-15 command buffers per frame. So maybe we should not pass the command encoder after all. Doing so is a very thing I added and it wasn't connected to a user ask

Reply:
I'll ask in wgpu matrix chat what folks recommend there :)

Reply:
there is this bit in favor of encouraging the user to pile in more stuff:
```
Don’t submit small command buffers. If a submission is processed on the GPU faster than new ones can be submitted on the CPU, it will result in wasted / idle GPU cycles.
```

Reply:
for reference, the same in DX12 flavor https://developer.nvidia.com/dx12-dos-and-donts#worksubmit
What I'm reading in all of this is that CommandBuffers should be [lagom](https://en.wikipedia.org/wiki/Lagom) ;)

Reply:
for posterity, comment from @cwfitzgerald (one of the primary maintainers of wgpu) on the topic:

> This is a bit of an unknown question for me. I did some primative testing a long long time ago on the submission performance relative to command buffer count and found it was nearly quadratic, but that was well before I was at in touch with the stack so I don't put too much value in that number
Note that wgpu splits each renderpass into its own command buffer and the cost of this is still up in the air (and part of why this question is important)